### PR TITLE
Parser

### PR DESCRIPTION
**Pull request by rachelriv**
I implemented a lot of fixes so we should have a very minimal parser working.

However, we still need the following: 
UPDATED: I think that one of the last things we need to implement now is parsing conditional expressions. We also need a lot of tests, including invalid test programs.

@alexschneider @juansc @whusted Please review and let me know what you think! Thanks guys :tea:

### CODE REVIEW

File: en [hunk 1]
```diff
@@ -0,0 +1,31 @@
+Type = require './type'
```
Comment (alexschneider) [inline en]: What's this file for? It looks like it came from Iki - accidental commit?

File: package.json [hunk 2]
```diff
@@ -17,10 +17,10 @@
     "intercept-stdout": "0.0.2",
     "mocha": "^2.1.0",
     "mocha-lcov-reporter": "0.0.1",
+    "sinon": "latest",
```
Comment (alexschneider) [inline package.json]: I'd rather freeze sinon and sinon-chai to a specific version - we ran into an issue last time with two versions not being compatible.

File: parser/parser.coffee [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+  new VariableDeclaration id, exp
+
+parseVarAssig = ->
+  id = match 'ID'
+  match '='
+  exp = parseExpression()
+  new VariableAssignment id, exp
+
+parseConditional = ->
+  if at 'if'
```
Comment (alexschneider) [inline parser/parser.coffee]: It doesn't look like we're matching:

```
if <exp>:
  block
else:
  block
end

(or even if blocks)
```

File: parser/parser.coffee [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+  else
+    parseExp0()
+
+
+parseExp0 = ->
+  left = parseExp1()
+  while at 'or'
+    op = match()
+    right = parseExp1()
+    left = new BinaryExpression(op, left, right)
```
Comment (alexschneider) [inline parser/parser.coffee]: It seems pretty arbitrary when we use parens and when we don't - I'd rather be consistent throughout the entire program. Is there a reason for using parens here?

File: parser/parser.coffee [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+    op = match()
+    operand = parseExp6()
+    new UnaryExpression(op, operand)
+  else
+    parseExp6()
+
+parseListLiteral = ->
+  elements = []
+  match '['
+  elements.push(parseExpression()) unless at ']'
```
Comment (alexschneider) [inline parser/parser.coffee]: Parens here

**Comment by alexschneider** [comment]
I think the entities should be in the root directory of our compiler, not in the ./parser directory. The entities are used by more than the parser, they're used by the analyzer and possibly other files.

File: package.json (hunk 2 above)
Comment (whusted) [inline package.json]: Agreed, I've had that problem too

File: en (hunk 1 above)
Comment (rachelriv) [inline en]: Yup, my bad.

File: package.json (hunk 2 above)
Comment (rachelriv) [inline package.json]: That is what shrinkwrap is for.

File: parser/parser.coffee (hunk 3 above)
Comment (rachelriv) [inline parser/parser.coffee]: I just used your parser code for `if` statements. I didn't have the chance to add anything to it yet.

File: package.json (hunk 2 above)
Comment (alexschneider) [inline package.json]: All the same, I'd rather have version freezes in our package.json in addition to shrinkwrap.

**Comment by rachelriv** [comment]
Having the entities in the root level is fine by me.

File: parser/parser.coffee (hunk 4 above)
Comment (rachelriv) [inline parser/parser.coffee]: You're right. We should probably just use parens when we don't have args and omit them when we do. I'll clean that up.

Comment (whusted) [inline parser/parser.coffee]: The parens here are probably residue from Iki's parser. But yeah, parens only when we have to.

**Pull request merged**